rebuild/src/public_key.cc

```diff
@@ -0,0 +1,154 @@
+        InstanceMethod(
+            "keyValidate",
+            &PublicKey::KeyValidate,
+            static_cast<napi_property_attributes>(napi_enumerable)),
+        InstanceMethod(
+            "isInfinity",
+            &PublicKey::IsInfinity,
+            static_cast<napi_property_attributes>(napi_enumerable)),
+        InstanceMethod(
+            "multiplyBy",
+            &PublicKey::MultiplyBy,
+            static_cast<napi_property_attributes>(napi_enumerable)),
+    };
+
+    Napi::Function ctr = DefineClass(env, "PublicKey", proto, module);
+    module->_public_key_ctr = Napi::Persistent(ctr);
+    exports.Set(Napi::String::New(env, "PublicKey"), ctr);
+
+    Napi::Object js_exports = exports.Get("BLST_CONSTANTS").As<Napi::Object>();
+    js_exports.Set(
+        Napi::String::New(env, "PUBLIC_KEY_LENGTH_COMPRESSED"),
+        Napi::Number::New(env, BLST_TS_PUBLIC_KEY_LENGTH_COMPRESSED));
+    js_exports.Set(
+        Napi::String::New(env, "PUBLIC_KEY_LENGTH_UNCOMPRESSED"),
+        Napi::Number::New(env, BLST_TS_PUBLIC_KEY_LENGTH_UNCOMPRESSED));
+}
+
+Napi::Value PublicKey::Deserialize(const Napi::CallbackInfo &info) {
+    BLST_TS_FUNCTION_PREAMBLE(info, env, module)
+    Napi::Value pk_bytes_val = info[0];
+    BLST_TS_UNWRAP_UINT_8_ARRAY(pk_bytes_val, pk_bytes, "pkBytes")
+
+    std::string err_out{"BLST_ERROR: pkBytes"};
+    if (!is_valid_length(
+            err_out,
+            pk_bytes.ByteLength(),
+            BLST_TS_PUBLIC_KEY_LENGTH_COMPRESSED,
+            BLST_TS_PUBLIC_KEY_LENGTH_UNCOMPRESSED)) {
+        Napi::TypeError::New(env, err_out).ThrowAsJavaScriptException();
+        return env.Undefined();
+    }
+
+    bool is_affine = false;
+    // but figure out if request for affine
+    if (!info[1].IsUndefined()) {
+        Napi::Value type_val = info[1].As<Napi::Value>();
+        if (!type_val.IsNumber()) {
+            Napi::TypeError::New(
+                env, "BLST_ERROR: type must be of enum CoordType (number)")
+                .ThrowAsJavaScriptException();
+            return env.Undefined();
+        }
+        switch (type_val.As<Napi::Number>().Uint32Value()) {
+            case 0:
+                is_affine = true;
+                break;
+            case 1:
+                // is_affine defaults to false
+                break;
+            default:
+                Napi::TypeError::New(
+                    env,
+                    "BLST_ERROR: must be CoordType.affine or "
+                    "CoordType.jacobian")
+                    .ThrowAsJavaScriptException();
+                return env.Undefined();
+        }
+    }
+
+    // try/catch here because using untrusted bytes for point creation and
+    // blst library will throw for invalid points
+    try {
+        if (is_affine) {
+            return scope.Escape(
+                module->_public_key_ctr.New({Napi::External<P1Wrapper>::New(
+                    env,
+                    new P1Affine{blst::P1_Affine{
```

Comment:
Yes.  Napi::External can only accept a raw pointer.  It is why its passed all the way through to the unique_ptr in the P1Wrapper upon initialization in the constructor.  There is no way around this because the value passes through the Isolate (into and back out of JS context).  Its possible to pass a cleanup function to the External but It is eligible for garbage collection once the constructor returns and the value is no longer used.  I want the unique_ptr to own the object not the External.  I suppose I could keep the external around and hold that on the P1Wrapper but then there is an extra roundtrip to the Isolate to retrieve the data pointer and then a second indirection to retrieve the actual value and isolate lookups are SLOW by nature.  All `JSValue`s are really just addresses in the Isolate and the isolate holds the actual memory address for the target.